curvefs/src/metaserver/s3compact_wq_impl.cpp

```diff
@@ -139,112 +140,77 @@ void S3CompactWorkQueueImpl::DeleteObjs(const std::vector<std::string>& objs,
 
 std::list<struct S3CompactWorkQueueImpl::Node>
 S3CompactWorkQueueImpl::BuildValidList(const S3ChunkInfoList& s3chunkinfolist,
-                                       uint64_t inodeLen) {
-    std::list<struct S3CompactWorkQueueImpl::Node> validList;
-    std::function<void(struct Node)> addToValidList;
-    // always add newer s3chunkinfo to list
-    // [begin, end]
-    addToValidList = [&](struct Node newNode) {
-        // maybe truncated smaller
-        if (newNode.end >= inodeLen) newNode.end = inodeLen - 1;
-        for (auto it = validList.begin(); it != validList.end();) {
-            // merge from list head to tail
-            // try merge B to existing A
-            // data from B is newer than A
-            uint64_t nodeBegin = it->begin;
-            uint64_t nodeEnd = it->end;
-            if (newNode.end < nodeBegin) {
-                // A,B no overlap, B is left A, just add to list
-                validList.emplace(it, newNode);
-                return;
-            } else if (newNode.begin > nodeEnd) {
-                // A,B no overlap, B is right of A
-                // skip current node to next node
-                ++it;
-            } else if (newNode.begin <= nodeBegin && newNode.end >= nodeEnd) {
-                // B contains A, 1 part or split to 2 part
-                *it = newNode;
-                it->end = nodeEnd;
-                if (newNode.end > nodeEnd) {
-                    auto n = newNode;
-                    n.begin = nodeEnd + 1;
-                    addToValidList(std::move(n));
+                                       uint64_t inodeLen, uint64_t index,
+                                       uint64_t chunkSize) {
+    std::list<std::pair<uint64_t, uint64_t>> freeList;  // [begin, end]
+    freeList.emplace_back(chunkSize * index,
+                          std::min(chunkSize * (index + 1) - 1,
+                                   inodeLen - 1));  // start with full chunk
+    std::map<uint64_t, std::pair<uint64_t, uint64_t>>
+        used;  // begin -> pair(end, i)
+    auto fill = [&](uint64_t i) {
```

Comment:
it's better to make this a util function and do more testing

Reply:
current test cases have covered all branchs in this lambda function